docs/py_docs.rst

```diff
@@ -838,35 +1042,73 @@ At minimum, constants/properties/attributes should have a summary line, but can
 5. :ref:`References <py-docstring-references>` (optional)
 6. :ref:`Examples <py-docstring-examples>` (optional)
 
-In the short summary, a description of the type should be included:
+For example:
 
 .. code-block:: py
 
    NAME = 'LSST'
-   """Name of the project (str)"""
+   """Name of the project (`str`)."""
 
-Note that class attributes can alternatively be documented in an :ref:`Attributes <py-docstring-attributes>` section of the class's docstring.
-This is particularly useful when the attribute is not set in the class scope, but rather in a method such as ``__init__``.
+Note:
 
-.. code-block:: py
+- The docstring appears directly below the constant or class attribute.
+- The type is included in parentheses at the end of the summary line.
 
-   class Answer(object):
-       """Container for the answer.
-       
-       Attributes
-       ----------
-       answer : obj
-          The answer.
-       source
+In many classes, public attributes are set in the ``__init__`` method.
+The best way to document these public attributes is by declaring the attribute at the class level and including a docstring with that declaration:
+
+.. code-block:: python
+
+   class Metric(object):
+       """Verification metric.
+
+       [...]
        """
 
-       def __init__(self):
-           self.contents = 42
+       name = None
+       """Name of the metric (`str`)."""
+
+       unit = None
+       """Units of the metric (`astropy.units.Unit`)."""
+
+       def __init__(name, unit):
+           self.name = name
+           self.unit = unit
+
+Private attributes (prefixed by underscores: ``self._myAttribute``) do not need to be documented with docstrings.
+
+.. _py-docstring-property-structure:
+
+Documenting Class Properties
+============================
+
+Properties are documented like :ref:`class attributes <py-docstring-attribute-constants-structure>` rather than methods.
+After all, properties are designed to appear to the user like simple attributes.
+
+For example:
+
+.. code-block:: python
+
+   class Measurement(object):
+
+       # ...
 
        @property
-       def source(self):
-           """Purveyor of the answer."""
-           return 'Deep Thought'
+       def quantity(self):
+           """The measurement quantity (`astropy.units.Quantity`, mutable).
+           """
+           # ...
+
+       @quantity.setter
+       def quantity(self, q):
+           # ...
+
+Note:
+
+- Do not use the :ref:`Returns section <py-docstring-returns>` in the property's docstring.
+  Instead, include type information in the summary, as is done for :ref:`class attributes <py-docstring-attribute-constants-structure>`.
+- Along with type information, include whether the property is "immutable" or "mutable" (has a setter method).
```

Comment:
Can the user distinguish between attributes and properties in the generated documentation? Otherwise, it seems odd to make a property say "mutable" while not saying anything for an equivalent attribute.

Also, maybe Python usage of the term is different, but I would say that a property is unmodifiable or read-only, not immutable -- in general its value *can* change, just not by assignment through the property name.

Reply:
Good idea; I'll only have properties without setter say they are "read-only." That should fix the terminology and also add consistency with regular class attributes.